src/cursor.rs

```diff
@@ -90,32 +121,92 @@ impl Drop for Cursor {
             return;
         }
 
-        let namespace = self.get_more.namespace();
+        let namespace = self.get_more.namespace().clone();
+        let client = self.client.clone();
+        let cursor_id = self.get_more.cursor_id();
 
-        let _ = self.client.database(&namespace.db).run_command(
-            doc! {
-                "killCursors": &namespace.coll,
-                "cursors": [self.get_more.cursor_id()]
-            },
-            None,
-        );
+        RUNTIME.execute(async move {
+            let _: Result<_> = client.database(&namespace.db).run_command(
+                doc! {
+                    "killCursors": &namespace.coll,
+                    "cursors": [cursor_id]
+                },
+                None,
+            );
+        })
     }
 }
 
-impl Iterator for Cursor {
+impl Stream for Cursor {
     type Item = Result<Document>;
 
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.buffer.is_empty() && !self.exhausted {
-            match self.client.execute_operation(&self.get_more, None) {
-                Ok(get_more_result) => {
-                    self.buffer.extend(get_more_result.batch);
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        match self.state.take() {
+            // If the current state is Executing, then we check the progress of the getMore
+            // operation.
+            Some(PollState::Executing(mut future)) => match Pin::new(&mut future).poll(cx) {
+                // If the getMore is finished and successful, then we pop off the first document
+                // from the batch, set the poll state to Buffer, record whether the cursor is
+                // exhausted, and return the popped document.
+                Poll::Ready(Ok(get_more_result)) => {
+                    let mut buffer: VecDeque<_> = get_more_result.batch.into_iter().collect();
+                    let next_doc = buffer.pop_front();
+
+                    self.state = Some(PollState::Buffer(buffer));
                     self.exhausted = get_more_result.exhausted;
+                    Poll::Ready(next_doc.map(Ok))
```

Comment:
So this looks like we're doing the "tryNext" behavior here. After lots of discussion in Swift we ended up defaulting to the resending-getMore-on-empty-batch behavior for `next` while also offering a `tryNext` method if users wanted it. Do we plan on doing something similar here?

Reply:
Right now, we don't have tailable cursors, so there wouldn't be any user-facing difference regardless of the behavior we implement, correct? If so, we might as well wait until we implement tailable cursors to decide.

Reply:
We currently support tailable cursors via setting the cursor type on `FindOptions`. You can just keep calling `next` on the resulting cursor even after it returns an empty batch, since it'll never return `exhausted: true` from a getmore.

Reply:
Ah, true. Users generally won't be calling `poll_next` directly due to ergonomic reasons; it's mostly used under the hood in the `StreamExt` trait, where `next` is defined using it. If we implement `poll_next` to eagerly resend the getMore, I don't think there's any way we could define a higher-level alternative to `StreamExt::next` that doesn't eagerly send it. Since I wasn't involved in any of the Swift conversations, I think it would be helpful for me to ask you both about what the intended use cases are of `next` versus `tryNext` in Swift (although even if we do the same thing, we'll need to name them differently, since `StreamExt::try_next` is already a thing that means something different). Specifically, I'm not sure if an eagerly-getMore-sending `next` is intended to be used with the higher-level constructs that `StreamExt` provides (e.g. `map` and `collect`) or if it's just intended to be called directly. If the latter, I think it would make more sense to do the reverse of what Swift does and have `poll_next` not eagerly resend getMore, but if the former, we might need to do something more complex like having two separate cursors types (one eager and one lazy) and provide methods to convert them back and forth.

Reply:
To summarize offline conversation, we've elected to go with a two cursor strategy:
- Default `Cursor` implements the keep-resending-getMore-until-nonempty-batch behavior
- `Cursor` has a method on it that takes ownership and returns a new wrapping cursor (temporarily has type name `NonPatientCursor`) which does not resend getMore and instead returns empty batches.